File: apps/web/src/types/job.ts
/**
 * Machine Learning Job Orchestration TypeScript definitions.
 */

export type JobStatus =
  | 'PENDING'
  | 'QUEUED'
  | 'STARTING'
  | 'RUNNING'
  | 'VALIDATING'
  | 'TRAINING'
  | 'EVALUATING'
  | 'SAVING_MODEL'
  | 'COMPLETED'
  | 'FAILED'
  | 'CANCELLED'
  | 'RETRYING'

export interface TrainingOption {
  key: string
  display_name: string
}

export interface AlgorithmTrainingOption extends TrainingOption {
  task_type: 'classification' | 'regression'
}

export interface TrainingOptions {
  algorithms: AlgorithmTrainingOption[]
  scalers: TrainingOption[]
  imputers: TrainingOption[]
  default_cv_folds: number
  default_train_test_split: number
  min_train_test_split?: number
  max_train_test_split?: number
  min_cv_folds?: number
  max_cv_folds?: number
}

export const CANONICAL_TRAINING_OPTIONS: TrainingOptions = {
  algorithms: [
    { key: 'random_forest_classifier', display_name: 'Random Forest Classifier', task_type: 'classification' },
    { key: 'logistic_regression', display_name: 'Logistic Regression', task_type: 'classification' },
    { key: 'decision_tree_classifier', display_name: 'Decision Tree Classifier', task_type: 'classification' },
    { key: 'k_nearest_neighbors_classifier', display_name: 'K-Nearest Neighbors Classifier', task_type: 'classification' },
    { key: 'support_vector_classifier', display_name: 'Support Vector Classifier (SVC)', task_type: 'classification' },
    { key: 'gradient_boosting_classifier', display_name: 'Gradient Boosting Classifier', task_type: 'classification' },
    { key: 'xgboost_classifier', display_name: 'XGBoost Classifier', task_type: 'classification' },
    { key: 'lightgbm_classifier', display_name: 'LightGBM Classifier', task_type: 'classification' },
    { key: 'gaussian_nb', display_name: 'Naive Bayes (GaussianNB)', task_type: 'classification' },
    { key: 'ridge_classifier', display_name: 'Ridge Classifier', task_type: 'classification' },
    { key: 'random_forest_regressor', display_name: 'Random Forest Regressor', task_type: 'regression' },
    { key: 'linear_regression', display_name: 'Linear Regression', task_type: 'regression' },
    { key: 'decision_tree_regressor', display_name: 'Decision Tree Regressor', task_type: 'regression' },
    { key: 'k_nearest_neighbors_regressor', display_name: 'K-Nearest Neighbors Regressor', task_type: 'regression' },
    { key: 'support_vector_regressor', display_name: 'Support Vector Regressor (SVR)', task_type: 'regression' },
    { key: 'gradient_boosting_regressor', display_name: 'Gradient Boosting Regressor', task_type: 'regression' },
    { key: 'xgboost_regressor', display_name: 'XGBoost Regressor', task_type: 'regression' },
    { key: 'lightgbm_regressor', display_name: 'LightGBM Regressor', task_type: 'regression' },
    { key: 'ridge_regressor', display_name: 'Ridge Regressor', task_type: 'regression' },
    { key: 'lasso_regressor', display_name: 'Lasso Regressor', task_type: 'regression' },
  ],
  scalers: [
    { key: 'standard_scaler', display_name: 'StandardScaler' },
    { key: 'min_max_scaler', display_name: 'MinMaxScaler' },
    { key: 'robust_scaler', display_name: 'RobustScaler' },
    { key: 'max_abs_scaler', display_name: 'MaxAbsScaler' },
    { key: 'normalizer', display_name: 'Normalizer' },
  ],
  imputers: [
    { key: 'mean', display_name: 'Mean' },
    { key: 'median', display_name: 'Median' },
    { key: 'most_frequent', display_name: 'Most Frequent' },
    { key: 'constant', display_name: 'Constant' },
    { key: 'knn_imputer', display_name: 'K-Nearest Neighbors Imputer' },
  ],
  default_cv_folds: 5,
  default_train_test_split: 0.8,
  min_train_test_split: 0.5,
  max_train_test_split: 0.95,
  min_cv_folds: 2,
  max_cv_folds: 20,
}

export interface TrainingRequestPayload {
  dataset_id: string
  target_column: string
  feature_columns: string[]
  algorithm?: string
  scaler?: string
  imputer?: string
  train_test_split?: number
  random_seed?: number | null
  cross_validation?: number | null
  normalization?: boolean
  feature_selection?: string
  class_weight?: string
  notes?: string
  recommendation_job_id?: string | null
  selection_source?: 'recommended' | 'manual' | 'default'
}

export interface JobEntity {
  job_id: string
  dataset_id: string
  status: JobStatus
  created_at: string
  updated_at: string
  started_at?: string | null
  completed_at?: string | null
  cancelled_at?: string | null
  job_type: string
  algorithm: string
  target_column: string
  feature_columns: string[]
  progress: number
  current_stage: string
  message?: string | null
  estimated_seconds?: number | null
  worker_id?: string | null
  error_message?: string | null
  retry_count: number
  owner_id?: string | null
  metadata?: {
    scaler?: string
    imputer?: string
    random_seed?: number | null
    cross_validation?: number | null
    normalization?: boolean
    feature_selection?: string
    class_weight?: string
    recommendation_job_id?: string | null
    selection_source?: string
    [key: string]: unknown
  }
}

export interface JobListData {
  total: number
  jobs: JobEntity[]
}

export interface JobProgressInfo {
  job_id: string
  status: JobStatus
  progress: number
  current_stage: string
  message?: string | null
  estimated_seconds?: number | null
  estimated_seconds_remaining?: number | null
  error_message?: string | null
  updated_at: string
}

export interface JobCancelResult {
  job_id: string
  status: JobStatus
  message: string
}

export interface JobRetryResult {
  job_id: string
  status: JobStatus
  message: string
  retry_count: number
  new_job_id?: string | null
}

File: services/api/app/schemas/job.py
"""Pydantic schemas for Machine Learning Job Management and Training Configuration."""

import enum
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, field_validator, model_validator

from app.ml.algorithm_factory import ALGORITHM_REGISTRY
from app.ml.imputer_factory import IMPUTER_REGISTRY
from app.ml.scaler_factory import SCALER_REGISTRY


class JobStatusEnum(str, enum.Enum):
    """Deterministic Job Lifecycle Statuses."""

    PENDING = "PENDING"
    QUEUED = "QUEUED"
    STARTING = "STARTING"
    RUNNING = "RUNNING"
    VALIDATING = "VALIDATING"
    TRAINING = "TRAINING"
    EVALUATING = "EVALUATING"
    SAVING_MODEL = "SAVING_MODEL"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
    CANCELLED = "CANCELLED"
    RETRYING = "RETRYING"


class TrainingRequest(BaseModel):
    """Payload schema for POST /api/v1/jobs/train."""

    dataset_id: str = Field(..., description="ID of the uploaded dataset")
    target_column: str = Field(..., description="Target column variable name")
    feature_columns: list[str] = Field(..., description="List of feature column names")
    algorithm: str = Field("random_forest_classifier", description="Canonical ML model algorithm key")
    scaler: str | None = Field("standard_scaler", description="Canonical feature scaling strategy key")
    imputer: str | None = Field("median", description="Canonical missing-value imputation strategy key")
    train_test_split: float = Field(0.8, description="Train / Test split ratio (0.5 to 0.95)")
    random_seed: int | None = Field(42, description="Random seed for reproducibility")
    cross_validation: int | None = Field(5, description="Cross validation folds count")
    cv_n_splits: int | None = Field(None, description="Alias for cross validation folds count")
    normalization: bool | None = Field(True, description="Apply StandardScaler normalization (backward-compatible)")
    feature_selection: str | None = Field("all", description="Feature selection strategy")
    class_weight: str | None = Field("balanced", description="Class weighting mode")
    notes: str | None = Field("", description="Optional user notes")
    recommendation_job_id: str | None = Field(
        None, description="Originating recommendation benchmark job ID for provenance tracking"
    )
    selection_source: str = Field(
        "default",
        pattern=r"^(recommended|manual|default)$",
        description="Provenance of algorithm selection: 'recommended', 'manual', or 'default'",
    )

    @field_validator("algorithm")
    @classmethod
    def validate_algorithm(cls, algo: str) -> str:
        if algo not in ALGORITHM_REGISTRY:
            raise ValueError(f"Algorithm '{algo}' is not a published training option.")
        return algo

    @field_validator("scaler")
    @classmethod
    def validate_scaler(cls, scaler: str | None) -> str | None:
        if scaler is not None and scaler not in SCALER_REGISTRY:
            raise ValueError(f"Scaler '{scaler}' is not a published training option.")
        return scaler

    @field_validator("imputer")
    @classmethod
    def validate_imputer(cls, imputer: str | None) -> str | None:
        if imputer is not None and imputer not in IMPUTER_REGISTRY:
            raise ValueError(f"Imputer '{imputer}' is not a published training option.")
        return imputer

    @field_validator("train_test_split")
    @classmethod
    def validate_split(cls, split: float) -> float:
        if split < 0.5 or split > 0.95:
            raise ValueError("Train/Test split ratio must be between 0.50 and 0.95.")
        return split

    @model_validator(mode="after")
    def validate_features_and_target(self) -> "TrainingRequest":
        if self.feature_selection != "all" and (not self.feature_columns or len(self.feature_columns) == 0):
            raise ValueError("Feature list cannot be empty. Select at least one input feature.")
        if self.feature_columns and len(self.feature_columns) != len(set(self.feature_columns)):
            raise ValueError("Feature list contains duplicate column names.")
        if self.feature_columns and self.target_column in self.feature_columns:
            raise ValueError(
                f"Target column '{self.target_column}' cannot be included inside feature columns."
            )
        return self


class JobResponse(BaseModel):
    """Structured response schema representing an ML Job entity."""

    job_id: str
    dataset_id: str
    status: str
    created_at: datetime
    updated_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    cancelled_at: datetime | None = None
    job_type: str = "training"
    algorithm: str
    target_column: str
    feature_columns: list[str] = Field(default_factory=list)
    progress: float = 0.0
    current_stage: str = "Initialized"
    message: str | None = None
    estimated_seconds: float | None = 0.0
    worker_id: str | None = None
    error_message: str | None = None
    retry_count: int = 0
    owner_id: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)

    model_config = {"from_attributes": True}


class JobProgressResponse(BaseModel):
    """Real-time progress response schema for GET /api/v1/jobs/{job_id}/progress."""

    job_id: str
    status: str
    progress: float
    current_stage: str
    message: str | None = None
    estimated_seconds_remaining: float | None = 0.0


class JobListResponse(BaseModel):
    """Paginated list response for GET /api/v1/jobs."""

    total: int
    jobs: list[JobResponse] = Field(default_factory=list)


class JobCancelResponse(BaseModel):
    """Response schema for POST /api/v1/jobs/{job_id}/cancel."""

    job_id: str
    status: str = "CANCELLED"
    cancelled_at: datetime
    message: str


class JobRetryResponse(BaseModel):
    """Response schema for POST /api/v1/jobs/{job_id}/retry."""

    original_job_id: str
    new_job_id: str
    status: str = "QUEUED"
    retry_count: int
    message: str
